zlib-rs/src/inflate.rs

```diff
@@ -1592,22 +1603,46 @@ fn inflate_fast_help(state: &mut State, _start: usize) -> ReturnCode {
                                 panic!("INFLATE_ALLOW_INVALID_DISTANCE_TOOFAR_ARRR")
                             }
 
-                            // horrible reuse of this variable; find a good name
-                            let op = dist as usize - written;
-
-                            if state.window.next() == 0 {
-                                // very common case (TODO: why?)
-                                todo!()
-                            } else if state.window.next() >= op {
-                                let slice = &state.window.copy(op)[..Ord::min(op, len as usize)];
-                                writer.extend(slice);
+                            let mut op = dist as usize - written;
+                            let mut from;
+
+                            let window_next = state.window.next();
+
+                            if window_next == 0 {
+                                // This case is hit when the window has just wrapped around
+                                // by logic in `Window::extend`. It is special-cased because
+                                // apparently this is quite common.
+                                //
+                                // the match is at the end of the window, even though the next
+                                // position has now wrapped around.
+                                from = window_size - op;
+                            } else if window_next >= op {
+                                // the standard case: a contiguous copy from the window, no wrapping
+                                from = window_next - op;
                             } else {
-                                // wrap around in window
-                                todo!()
+                                // This case is hit when the window has recently wrapped around
+                                // by logic in `Window::extend`.
+                                //
+                                // The match is (partially) at the end of the window
+                                op -= window_next;
+                                from = window_size - op;
+
+                                if op < len as usize {
+                                    // This case is hit when part of the match is at the end of the
+                                    // window, and part of it has wrapped around to the start. Copy
+                                    // the end section here, the start section will be copied below.
+                                    len -= op as u16;
+                                    writer.extend(&state.window.as_slice()[from..][..op]);
+                                    from = 0;
+                                    op = window_next;
+                                }
                             }
```

Comment:
core achievement of this PR: we got rid of the `todo!`s here because we finally have test cases that hit all the branches!